# A framework for gene representation on spatial transcriptomics

## 1. Introduction
The framework incorporates ligand genes within the spatial niche into Transformer encoder of single-cell transcriptomics, and outputs gene embeddings specific to the intracellular context and spatial niche. These gene representations can serve as input of various downstream applications, including cell clustering, receptor-dependent gene network inference, interacting ligand-receptor pair identification, and in silico perturbation analysis of ligand-receptor interaction.

![Framework Architecture](https://github.com/csh3/stFormer/blob/main/schematic_overview.png)

The architecture is designed for ST data resolved at the single-cell level. We propose a biased cross-attention method to enable the framework to do learning with single-cell resolution on low-resolution, whole-transcriptome Visium data, which is generated by a widely used ST platform. 

The current version was pretrained on a combined corpus of two Visium datasets respectively from human developmental intestine and myocardial infarction heart tissue. The pretraining data totally comprises about 0.58 million spatially resolved cells. The framework is scalable to large pretraining corpus. 

## 2. Installation
You can install stFormer with the following command:

```
git clone https://github.com/csh3/stFormer.git
cd stFormer; pip install -e .
```

We recommend installing stFormer in a fresh python environment since we specify the installed versions of the python module dependencies.

## 3. Usage
The project consists of the following four folders.

* **stformer** folder holds source codes of the model.

* **data** folder holds the ST data for pretraining and downstream tasks. A download link for data is provided in the file *download.txt*.

* **pretraining** folder holds the pretrained model weights and pretraining script. A download link for model weights is provided in the file *download.txt*.

* **tasks** folder holds the jupyter notebooks and gene lists for performing downstream tasks, as well as the output results. It also holds the **scfoundation** folder, which contains files for the single-cell transcriptomics foundation model, scFoundation, for comparison.

If performing pretraining, you should move the python script to the main directory first using command       
`mv pretraining/pretraining.py .`

## 4. Reference
Cao, S. & Yuan, Y. A framework for gene representation on spatial transcriptomics. bioRxiv, 2024.2009.2027.615337 (2024). https://doi.org/10.1101/2024.09.27.615337



